### PR TITLE
Added deprecation note for the use of cluster name in path.data

### PR DESCRIPTION
**Pull request by turbo-ele**
This PR adds **release notes** for 2.0.0, where the effective data storage path was changed and adds a **cluster node check** for that, since this is going to be a breaking change in future versions of CrateDB. Currently there's a fallback which still accepts the old path structure.

This needs to be backported

### CODE REVIEW

File: blackbox/docs/release_notes/2.0.0.txt [hunk 1]
```diff
@@ -299,3 +302,28 @@ sets both the minimum and maximum memory to allocate, and should be set to
 whatever your previous ``CRATE_MAX_MEM`` was set to.
 
 .. _enterprise edition: https://crate.io/enterprise-edition/
+
+Cluster name in path data
+-------------------------
+
+In previous versions, the cluster name was a folder inside the ``path.data``
+directory. Now, the data is written directly in the ``path.data`` directory.
+If it still contains a folder equals the cluster name it is read at startup.
```
Comment (Autophagy) [inline blackbox/docs/release_notes/2.0.0.txt]: I would rephrase this to something like 'If the directory still contains a folder whose name is the same as the cluster name, it is read at start up.'

File: blackbox/docs/release_notes/2.0.0.txt [hunk 2]
```diff
@@ -299,3 +302,29 @@ sets both the minimum and maximum memory to allocate, and should be set to
...
+$PATH_DATA_DIR/$CLUSTER_NAME/nodes
+```
+
+needs to be changed to
+
+```
+$PATH_DATA_DIR/nodes
+```
+
+.. NOTE::
```
Comment (matriv) [inline blackbox/docs/release_notes/2.0.0.txt]: I would change it to WARNING

File: blackbox/docs/release_notes/2.0.0.txt [hunk 3]
```diff
@@ -299,3 +302,29 @@ sets both the minimum and maximum memory to allocate, and should be set to
...
+If the directory still contains a folder whose name is the same as the cluster
+name, it is read at startup. This behaviour is deprecated and will be removed
+in future versions of CrateDB. To resolve this, you will need to move node
+data one level upwards. E.g.:
+
+```
+$PATH_DATA_DIR/$CLUSTER_NAME/nodes
+```
+
+needs to be changed to
```
Comment (matriv) [inline blackbox/docs/release_notes/2.0.0.txt]: needs to be moved to

**Comment by chaudum** [comment]
deprecation **note** (not node) - both in the PR title and commit message

File: blackbox/docs/release_notes/2.0.0.txt [hunk 4]
```diff
@@ -299,3 +302,19 @@ sets both the minimum and maximum memory to allocate, and should be set to
...
+
+Cluster name in path data
+-------------------------
+
+The computation of the effective data directory path has changed in a way that
+the cluster name is not part of the path anymore. In previous versions it was
+``$PATH_DATA_DIR/$CLUSTER_NAME/nodes/`` and now it is
+``$PATH_DATA_DIR/nodes/``. This requires that the data directory is either
+moved to the new location or the ``path.data`` setting gets changed to point to
+the old location by appending the clustername to it. (e.g ``/data/`` gets
```
Comment (Autophagy) [inline blackbox/docs/release_notes/2.0.0.txt]: ~~gets~~ becomes

File: blackbox/docs/release_notes/2.0.0.txt [hunk 5]
```diff
@@ -299,3 +302,19 @@ sets both the minimum and maximum memory to allocate, and should be set to
 whatever your previous ``CRATE_MAX_MEM`` was set to.
 
 .. _enterprise edition: https://crate.io/enterprise-edition/
+
+Cluster name in path data
+-------------------------
+
+The computation of the effective data directory path has changed in a way that
```
Comment (mfussenegger) [inline blackbox/docs/release_notes/2.0.0.txt]: This doesn't seem to be correct. If I run Crate with 1.1 with a custom data.path and upgrade to 2.0 without changing the data.path setting it still picks up the data.

File: sql/src/main/java/io/crate/operation/reference/sys/check/node/ClusterNameInPathDataNodesSysCheck.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+     *
+     * @param clusterName   current cluster name
+     * @param pathDataPaths all paths specified in path.data
+     * @param nodeDataPaths all paths specified in NodeEnvironment
+     *
+     * @return list of all paths that contain the cluster name as folder
+     */
+    @VisibleForTesting
+    static List<String> comparePathDataWithNodeDataPaths(String clusterName,
+                                                         List<String> pathDataPaths,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/reference/sys/check/node/ClusterNameInPathDataNodesSysCheck.java]: I'd rename to parameters to `configuredDataPaths` and `resolvedDataPaths`

File: sql/src/main/java/io/crate/operation/reference/sys/check/node/ClusterNameInPathDataNodesSysCheck.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+        return result;
+    }
+
+    /**
+     * Creates list of strings from NodeEnvironment paths and cuts off nodes folder.
+     *
+     * @return list of node data paths without nodes suffix
+     */
+    private static List<String> convertNodeDataPaths(Path[] paths) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/reference/sys/check/node/ClusterNameInPathDataNodesSysCheck.java]: maybe rename the method to `trimNodesSuffix`

File: sql/src/main/java/io/crate/operation/reference/sys/check/node/SysNodeChecksModule.java [hunk 8]
```diff
@@ -33,10 +33,14 @@ protected void configure() {
...
-        b.addBinding(LowDiskWatermarkNodesSysCheck.ID).to(LowDiskWatermarkNodesSysCheck.class);
+        // Node checks ordered by ID. New ID must be max ID + 1 and must not be reused.
+        b.addBinding(RecoveryExpectedNodesSysCheck.ID).to(RecoveryExpectedNodesSysCheck.class);
         b.addBinding(RecoveryAfterNodesSysCheck.ID).to(RecoveryAfterNodesSysCheck.class);
         b.addBinding(RecoveryAfterTimeSysCheck.ID).to(RecoveryAfterTimeSysCheck.class);
-        b.addBinding(RecoveryExpectedNodesSysCheck.ID).to(RecoveryExpectedNodesSysCheck.class);
+        b.addBinding(HighDiskWatermarkNodesSysCheck.ID).to(HighDiskWatermarkNodesSysCheck.class);
+        b.addBinding(LowDiskWatermarkNodesSysCheck.ID).to(LowDiskWatermarkNodesSysCheck.class);
+
+        // Temporary checks. IDs beginning with 1000.
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/reference/sys/check/node/SysNodeChecksModule.java]: Not sure if this is necessary? This implies that the IDs < 1000 are guaranteed to stay, which might not be the case.

File: sql/src/test/java/io/crate/integrationtests/SysNodeCheckerIntegrationTest.java [hunk 9]
```diff
@@ -36,10 +45,28 @@
 @UseJdbc
 public class SysNodeCheckerIntegrationTest extends SQLTransportIntegrationTest {
 
+    private static List<String> dirs = new ArrayList<>();
+
+    @Override
+    // TODO: This is for testing the ClusterNameInPathDataNodesSysCheck and
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SysNodeCheckerIntegrationTest.java]: Maybe rename all TODOs in this PR to ES6-TODO or something like that. We've already a lot of other TODOS and so it will at least be easier to grep for these specifically

File: sql/src/test/java/io/crate/integrationtests/SysNodeCheckerIntegrationTest.java [hunk 10]
```diff
@@ -78,4 +107,20 @@ public void testUpdateAcknowledgedFromReference() throws Exception {
...
+     * contains cluster name folder.
+     *
+     * TODO: Remove this when ES >= 6.0
+     */
+    @Test
+    public void testClusterNameInPathDataNodesSysCheck() throws Exception {
+        execute("select description, passed from sys.node_checks where not passed and id=1000");
+        assertThat(response.rowCount(), is(2L));
+        String description = (String) response.rows()[0][0];
+        assertTrue(description.contains(dirs.get(0)) || description.contains(dirs.get(1)));
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SysNodeCheckerIntegrationTest.java]: We generally prefer the `assertThat(..., is()` pattern. Or in this case `assertThat(description, contains(..))`  - this will result in better error messages.

File: sql/src/test/java/io/crate/integrationtests/SysNodeResiliencyIntegrationTest.java [hunk 11]
```diff
@@ -77,4 +76,16 @@ public void testTimingOutNode() throws Exception {
...
     }
+
+    /**
+     * Test cluster name check, when path.data does not contain cluster
+     * name as a folder.
+     *
+     * TODO: Remove this when ES >= 6.0
+     */
+    @Test
+    public void testClusterNameInPathCheck() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SysNodeResiliencyIntegrationTest.java]: maybe make the test method more descriptive - e.g.  `testClusterNameInPathCheckDoesNotFailWithDefaultConfig`

File: blackbox/docs/release_notes/2.0.0.txt (hunk 5 above)
Comment (turbo-ele) [inline blackbox/docs/release_notes/2.0.0.txt]: This is because there's a fallback for that, which i'm describing in the last paragraph. Probably i should write
`This fallback will be removed in future versions of CrateDB` instead ?

Comment (mfussenegger) [inline blackbox/docs/release_notes/2.0.0.txt]: Not sure how to phrase it. But there shouldn't be any contradiction in the text.

I think it would also be okay to just add in the changes.txt a new entry about the new cluster check (basically a "Added deprecation notices to clusterName usage in `data.path`")

I don't think that people who are already on 2.1 go back to read the 2.0 release notes.

Comment (turbo-ele) [inline blackbox/docs/release_notes/2.0.0.txt]: I just talked to @joemoe about that. Since this currently is 'just' deprecated, there is no need to add it to the changes, or current release notes:
1. We added a cluster check, that informs the user and is backported to 2.1
2. When we upgrade to ES >= 6.0 which will cause the fallback to be removed, another note about this should be added to the release notes

File: sql/src/test/java/io/crate/integrationtests/SysNodeResiliencyIntegrationTest.java [hunk 12]
```diff
@@ -77,4 +76,16 @@ public void testTimingOutNode() throws Exception {
...
     }
+
+    /**
+     * Test cluster name check, when path.data does not contain cluster
+     * name as a folder.
+     *
+     * TODO-ES6: Remove this when ES >= 6.0
+     */
+    @Test
+    public void testClusterNameInPathCheckDoesNotFailWithDefaultConfig() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SysNodeResiliencyIntegrationTest.java]: btw this test is actually a bit missleading - it tests that the default test-setup contains a non-deprecated data path. But doesn't really verify that our regular default config is okay.

File: blackbox/docs/release_notes/2.0.0.txt (hunk 5 above)
Comment (mfussenegger) [inline blackbox/docs/release_notes/2.0.0.txt]: I still think that the phrasing here could be improved a bit. It reads like "the cluster name can no longer be part of the data.path directory, but there is a fallback so it still can be part of it"

File: sql/src/test/java/io/crate/integrationtests/SysNodeResiliencyIntegrationTest.java (hunk 12 above)
Comment (turbo-ele) [inline sql/src/test/java/io/crate/integrationtests/SysNodeResiliencyIntegrationTest.java]: i'm going to add an extra assertion that cluster name folder really doesn't exist.

How would you like to verify that the default config is okay? And should this be part of this test case?

Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SysNodeResiliencyIntegrationTest.java]: I think it's okay as there are unit test - I was just questioning if this integration test even makes sense or if it could be removed. But I leave it up to you
